README

	T3X/0 Extended Edition - A New T3X Compiler
	===========================================

	T3X is a family of small, portable, procedural, block-structured,
	recursive, and almost typeless programming languages. The T3X
	syntax is similar to Pascal, its semantics resembles BCPL's.

	This Edition includes new commands and operations aliases.

	T3X/0 is the latest member of the T3X family. Its design is
	based on the experience made with its predecessors. It is
	simpler than the full (Release 7) language, but more complete
	than the minimalistic T3X9 branch.

	The compiler presented here is a single-stage two-pass compiler
	that is aimed to be ported quickly to new platforms, typically
	by just writing a new machine description file. The compiler
	can run on systems as minimal as CP/M on a Z80 CPU and easily
	fits in a tiny model DOS executable. It self-compiles in ten
	minutes on a 4MHz CP/M system (and in the blink of an eye on a
	modern Unix system).


	*** Currently Supported Platforms ***

	Operating System   CPU    Output format       Requires
	Generic Unix       VM     Tcode/0
	Generic Unix       386    assembly language   C compiler
	FreeBSD            386    assembly language   assembler, loader
	CP/M 2.2           Z80    COM                 48K bytes TPA
	PCDOS 2.0          8086   COM

	The compiler can operate natively or as a cross-compiler. Almost
	all combinations are supported: for example, you can cross-
	compile on CP/M for the TCVM, on the TVCM for Unix, on Unix for
	DOS, and on DOS for CP/M. Or on CP/M for DOS, etc. The only
	combinations that are not possible are cross-compilation to
	Unix on CP/M or DOS, because the source code emitter is too
	large.

	NOTE: when compilation is aborted and re-compilation fails,
	this is usually due to the wrong code generator being linked
	in place. In this case the following line should bring the
	code back to a useful state:

	   make clean; make reset; make


	*** Quick Installation ***

	If you just want to install T3X/0, edit the T3XDIR, BINDIR, and
	HOST variables in Makefile and then run "make install". Use
	HOST=unix, even on FreeBSD, unless you want the static code
	backend (with does not support T.BREAK, though).

	If you are not using a 386 processor (or a x86-64 processor with
	32-bit compatibility support), you will have to use the bytecode
	compiler. Just running "make" will build it.


	*** Bootstrapping ***

	On a Unix system, just run "make".

	The T3X/0 archive contains a pre-compiled Tcode/0 executable of
	the final T3X/0->Tcode/0 compiler as well as a bootstrapping
	compiler in T3Xr7. So bootstrapping can be accomplished in two
	ways:

	(1) by using the pre-compiled binary BIN/TXTRN0.TC to
	    self-compile.

	    $ cc -m32 -o tcvm tcvm.c
	    $ ./tcvm bin/txtrn0.tc txtrn

	(2) or, by compiling the boostrapping compiler TXTRN0.T with
	    T3Xr7 and then compiling the final compiler with that.

	    $ tx txtrn0.t
	    $ txx txtrn0 txtrn

	In either case an existing C compiler is needed to compile
	the Tcode/0 Virtual Machine (TCVM).

	$ cc -o -m32 tcvm tcvm.c

	The TCVM is a very basic C program with a length of less than
	250 lines. It is a 32-bit program, though, and needs more than
	64K bytes of data. You *might* get away with large model on DOS.
	On 64-bit platforms it must be compiled in 32-bit mode (-m32).

	There are also pre-compiled binaries for DOS and CP/M in the
	BIN directory, in case you should want to bootstrap on such
	a system. See below section for further details.


	*** Testing ***

	On a Unix system, just run "make test" and "make triple".

	There is a simple test suite in PROGRAMS/TEST.T. The compiler
	should at least pass these tests. To perform the tests, compile
	the program and run it:

	$ tcvm txtrn.tc programs/test test
	$ tcvm test.tc

	The output of the program should be self-explanatory.

	Then you can also run the triple test. First re-compile the
	compiler with itself:

	$ tcvm txtrn.tc txtrn txtrn1

	and then re-compile it again using the compiler generated in
	the previous step:

	$ tcvm txtrn1.tc txtrn txtrn2

	There should not be any differences between the generated
	compilers:

	$ cmp txtrn1.tc txtrn2.tc


	*** Usage ***

	The (interpreted) T3X/0 compiler is invoked by using the command
	line

	tcvm txtrn.tc program

	The compiler will then compile the file "program.t" and output
	a TCVM binary named "program.tc". The compiled program can then
	itself be run by the TCVM:

	tcvm program.tc

	When a second file name is passed to the compiler, it will
	write the compiled program to that file.

	tcvm program foo

	will compile "program.t" to "foo.tc".

	When the second parameter of the compiler is the string /v (or
	/V), then it will compile to "program.tc" and in addition echo
	the name of every function definition while compiling. This can
	be entertaining on slow systems.

	When a native code binary for a compiler exists, the compiler
	can be invoked by just typing "txtrn" instead of "tcvm txtrn.tc".

	Some compilers generate assembly language output, which needs
	to be further compiled and linked. The compiler driver script
	BIN/TX0.SH takes care of this. The details are outlined below.


	*** Compiling the Compiler ***

	------------------------------------------------------------
	!!! This process is automated by the BIN/BUILD.SH script !!!
	------------------------------------------------------------

	First select a platform to compile for. The target files are
	named TXname (the machine-specific T3X core module) and CGname
	(the code generator). The target files for the Tcode/0 machine
	can be found in the root directory of the T3X source code tree
	and the other targets are in the "targets" directory.

	Then there are two different emitters, one for direct output
	of binary programs and one for source-to-source compilation.

	To cross-compile the compiler for a system other than the
	Tcode/0 virtual machine (TCVM), follow these steps:

	1) Select an emitter: the DOS and CP/M ports need the binary
	   emitter "txemtbin.t" and the Unix ports needs the source
	   code emitter "txemtsrc.t". Link or copy the needed emitter
	   to "txemit.t".

	   E.g.: ln -fs txemtbin.t txemit.t

	2) Link the file CGname to "cg.t". Leave "t3x.t" linked to the
	   Tcode machine backend ("txtcvm.t").

	   E.g.: ln -fs targets/cgcpmz80.t cg.t
	         ln -fs txtcvm.t t3x.t  # this should not be necessary

	3) Edit the file "cg.t" and change the value of the modpath()
	   function to reflect your installation. Any trailing path
	   seperators must be given (e.g. "lib/" instead of "lib").
	   Note that a prefix of "" means leave the input file as is.
	   On CP/M only drive letters (A:, etc) may be specified.

	4) Compile the compiler. This step will create a TCVM binary
	   "tx-name0" that generates output for the selected platform.

	   E.g.: tcvm txtrn.tc txtrn tx-cpm0

	5) Link the file TXname to "t3x.t".

	   E.g.: ln -fs targets/txcpmz80.t t3x.t

	6) Compile the compiler again, this time using the compiler
	   created in the previous step. This step will create a binary
	   for the selected platform.

	   E.g.: tcvm tx-cpm0.tc txtrn tx-cpm

	   In the example the final compiler will be named "tx-cpm.com".
	   It will run on CP/M and generate native code for CP/M.

	7) Remove the cross-compiler tx-name0.tc and reset "t3x.t" and
	   "cg.t" to their TCVM targets, "txtcvm.t" and "cgtcvm.t".
	   Also reset "txemit.t" to "txemitbin.t".

	The resulting binary can then be used to compile T3X programs
	(including the compiler itself) on the selected host platform.
	All that is needed is the binary itself and the corresponding
	TXname file (renamed "t3x.t"). It might be necessary to convert
	the t3x.t file to local text file convention (e.g. on CP/M).

	All ports that use the source code emitter (like the Unix ports)
	emit assembly language, which must be further compiled and
	linked with a system-specific assembler and linker. For the
	FreeBSD-386 port, the resulting assembly language file can
	simply be passed to the C compiler controller, cc. E.g.:

	   tcvm tx-fbsd.tc txtrn tx-fbsd

	will generate a file named "tx-fbsd.s", which can then be
	compiled with cc:

	   cc -m32 -o tx-fbsd tx-fbsd.s

	(The -m32 option can/must be omitted on 32-bit machines.)

	The portable Unix backend works in a similar way, but also needs
	a runtime library. To finish compilation of the portable Unix
	version of the compiler, run

	   cc -m32 -o tx-unix tx-unix.s targets/ux386lib.c


	*** Building other Cross Compilers ***

	There is a total of 14 compilers that can be built from the
	T3X/0 code: four host platforms and four target platforms make
	a total of 16, but two combinations are not possible, because
	the compiler would be too large to fit in the resulting binary.
	These compilers exist:

	Target --> TCVM        CP/M        DOS        Unix*
	Host
	----
	TCVM       txtrn.tc    tx-cpm.tc   tx-dos.tc  tx-unix.tc

	CP/M       tx-tcvm.com tx-cpm.com  tx-dos.com n/a

	DOS        tx-tcvm.com tx-cpm.com  tx-dos.com n/a

	Unix*      tx-tcvm     tx-cpm      tx-dos     tx-unix

	* There is a FreeBSD backend named "fbsd" that can be used in
	  the place of the "unix" backend on FreeBSD-386. It uses a
	  runtime wrapper in assembly language instead of C. Note that
	  this backend does NOT offer a T.BREAK procedure, though.

	The BUILD.SH script in the "bin" directory can be used to
	build any of the above compilers by just specifying the host
	and target platforms as parameters. For instance,

	$ bin/build.sh cpm dos

	will generate an executable named "tx-dos.com" that will run
	on CP/M and generate code for DOS. Note that the DOS compiler
	for generating DOS output will have the same name, so when you
	plan to build both, you will have to rename one of them.

	Also note that when you have both a native compiler and a cross
	compiler on CP/M, they will need different "t3x.t" files! It
	is up to you to invent a workable solution.


	*** Using the Cross Compilers ***

	The "install" target in the Makefile will build native (cross)
	compilers for all supported platforms. I.e. the compilers will
	run on Unix-386 and emit code for the TCVM and for DOS-8086,
	CP/M-Z80, and Unix-386 / FreeBSD-386.

	The TX0 script will take care of all necessary steps during
	compilation. For instance

	$ tx0 foo

	will compile foo.t to a Unix-386 executable named "foo", and

	$ tx0 -t dos foo

	will compile foo.t to a DOS-8086 executable named "foo.com".


	*** Using the T3X/0 Compiler on DOS and CP/M ***

	There are pre-compiled compilers for DOS and CP/M in the BIN
	directory. To install a T3X/0 compiler on CP/M, copy the file
	BIN/TX-CPM.COM to TX0.COM on your CP/M disk. Then convert the
	file CG/TXCPMZ80.T to CP/M text file format and install it as
	T3X.T on your CP/M disk. The program PROGRAMS/CPMFILE.T can be
	used to convert the file.

	$ tcvm txtrn.tc programs/cpmfile cpmfile
	$ cp targets/txcpmz80.t t3x-cpm.t
	$ tcvm cpmfile.tc t3x-cpm.t /R
	$ # copy t3x-cpm.t to your CP/M system as T3X.T

	To install the compiler on DOS, copy the file BIN/TX-DOS.COM
	to TX0.COM on your DOS disk and install CG/TXDOS86C.T as T3X.T.
	Conversion to DOS text file format is optional (it can be done
	with the program PROGRAMS/DOSFILE.T).

	If you want to bootstrap the compiler on CP/M, you will also
	need the file CG/CGCPMZ80.T as CG.T, TXEMTBIN.T as TXEMIT.T,
	and TXTRN.T. All these files have to be converted to CP/M text
	file format.

	On DOS you need CG/CGDOS86C.T as CG.T and TXEMTBIN.T as
	TXEMIT.T as well as TXTRN.T. Conversion to DOS text format
	is optional.

	To compile the compiler on either system, run

	TX0 TXTRN

	which will result in a new compiler named TXTRN.COM. On CP/M
	(@ 4MHz) this will take about 10 minutes. You can use the /V
	option to entertain yourself while the compiler compiles.



CHANGES

	2023-02-13

	- Added the T.INTR86 procedure and T3X.REG86 structure to the
	  DOS runtime module.
	- Implemented T.TRUNC in FreeBSD runtime module.


	2023-02-12

	- Implemented T.TRUNC and T.BREAK in DOS runtime module.


	2023-01-17

	- Implemented T.TRUNC (truncate files) and T.BREAK (intercept
	  keyboard break) in Unix runtime module.


	2022-12-31

	- Made -C (where C is a constant) a valid constant expression.


	2022-12-10

	- Added "-e external_module" option to TX0.
	  Try "tx0 -e signal.c mu" in  programs/.


	2022-12-06

	- Renamed UNX386 backend FBD386 (FreeBSD-386).
	- bin/build and bin/tx0 now use target 'fbsd' instead of 'unix'.
	- Added more generic Unix-386 backend (UNX386).


	2022-10-22

	- Fix: f(p) miscompiled, if P is a procedure.
	- Added shellsrt module


	2022-10-18

	- Unix-386 T.NEWLINE returned wrong sequence.
	- Fixed some errors in the manual (t3x.txt).


	Differences between T3Xr7 and T3X/0

	Added string literals as abbreviations to packed vectors.

	Changed module syntax:
	MODULE name; declarations [compound] END
	If a compound statement is present, it will execute when
	the module is USEd for the first time, or at the beginning
	of program execution, if the module declaration is not in
	a separate file.

	PUBLIC declarations in modules can be imported.

	Changed module import syntax:
	USE name;
	USE name: alias;
	USE name: name;
	Will load module from sources, but only when not already
	present. Using alias=name overrides internal module names.

	Removed classes and objects.

	Constant expressions limited to *, +, |.

	Added EXTERN procedures for ports with source output:
	EXTERN name(arity), ... ;

	Added INLINE procedures for ports with binary output:
	INLINE name(arity) = [ byte, ... ], ... ;

	Jumps out of blocks allocating local storage (LEAVE, LOOP)
	are safe. Storage will be deallocated before exiting.

	Full T3X core library supported, including the functions BPW,
	NEWLINE, GETARG, CREATE, SEEK, REMOVE, RENAME, as well as
	OAPPND mode. (SEEK and OAPPND are not supported on CP/M.)



tcvm.c
/*
 * Ad-hoc 16-bit Tcode/0 virtual machine
 * Nils M Holm, 2017,2022,2023 0BSD license
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>

#undef DEBUG

/*
 * Define EXTRA to enable t3x.trunc and t3x.break.
 */
#define EXTRA

#include <signal.h>

#define MEMSIZE	0xfffe

#define byte	unsigned char
#define sbyte	signed char
#define cell	int

byte	*M;
cell	Red;

char	**Args;
int	Narg;

void writes(char *s) { write(1, s, strlen(s)); }
void wlog(char *s) { write(2, s, strlen(s)); }

void fail(char *s) {
	wlog("tcvm: ");
	wlog(s);
	wlog("\n");
	exit(1);
}

void load(char *s) {
	int	fd;
	cell	tp;
	char	b[6];
	char	p[100];

	fd = open(s, O_RDONLY);
	if (fd < 0 && strlen(s) < 97) {
		strcpy(p, s);
		strcat(p, ".tc");
		fd = open(p, O_RDONLY);
	}
	if (fd < 0) fail("could not open program");
	read(fd, b, 6);
	if (memcmp(b, "T3X0", 4)) fail("not a tcvm program");
	tp = b[4] | (b[5] << 8);
	M = malloc(MEMSIZE);
	if (NULL == M) fail("not enough memory");
	Red = read(fd, M, MEMSIZE);
	close(fd);
}

cell	A, F, I, P;

cell w(cell a) { return  M[a+0] | (M[a+1] << 8); }

void Sw(cell a, cell w) {
	M[a+0] =  w        & 255;
	M[a+1] = (w >> 8)  & 255;
}

void push(cell x) {
	P -= 2;
	Sw(P, x);
}

cell pop(void) {
	P += 2;
	return w(P-2);
}

cell S(int x) { return x > 32767? x-65536: x; }

#define a()	w(I+1)
#define a2()	w(I+3)

cell memscan(cell p, cell c, cell k) {
	cell	i;

	for (i=0; i<k; i++)
		if (M[p+i] == c)
			return i;
	return 0xffff;
}

int getarg(int n, char *s, int k) {
	int	j, m;

	n++;
	k--;
	if (n >= Narg) return -1;
	j = strlen(Args[n]);
	m = j>=k? k: j;
	memcpy(s, Args[n], m);
	s[m] = 0;
	return m;
}

cell t3xopen(char *s, int mode) {
	int	r;

	if (1 == mode)
		r = creat((char *) &M[w(P+4)], 0644);
	else if (3 == mode)
		r = open((char *) &M[w(P+4)], O_WRONLY);
	else
		r = open((char *) &M[w(P+4)], w(P+2));
	if (3 == mode) lseek(r, 0L, SEEK_END);
	return r;
}

cell t3xseek(cell fd, unsigned cell where, cell how) {
	long	w;
	int	h;

	switch (how) {
		case 0:	w = where; h = SEEK_SET; break;
		case 1: w = where; h = SEEK_CUR; break;
		case 2: w = where; w = -w; h = SEEK_END; break;
		case 3: w = where; w = -w; h = SEEK_CUR; break;
		default: return 0xffff;
	}
	return lseek(fd, w, h) < 0? 0xffff: 0;
}

cell t3xtrunc(cell fd) {
#ifndef EXTRA
	fail("t3x.trunc not implemented");
	return 0;
#else
 #ifdef unix
	return ftruncate(fd, lseek(fd, 0, SEEK_CUR));
 #endif
#endif
}

char	*Sem;

void handle_break(int dummy) {
	Sem[0] = Sem[1] = -1;
#ifdef unix
	signal(SIGINT, handle_break);
#endif
}

cell t3xbreak(cell sem) {
#ifndef EXTRA
	fail("t3x.break not implemented");
#else
 #ifdef unix
	if (0 == sem) {
		signal(SIGINT, SIG_DFL);
	}
	else if (1 == sem) {
		/* ignore */
	}
	else {
		Sem = (char *) &M[sem];
		Sem[0] = Sem[1] = 0;
		signal(SIGINT, handle_break);
	}
 #endif
#endif
	return 0;
}

cell libcall(cell n) {
	cell	r;

#ifdef DEBUG
	printf("LIBCALL(%d): %x %x %x %x\n", n, w(P+6), w(P+4), w(P+2), w(P));
#endif
	switch (n) {
	case  0: r = 2; break;
	case  1: strcpy((char *) &M[w(P+2)], "\n"); r = w(P+2); break;
	case  2: r = memcmp(&M[w(P+6)], &M[w(P+4)], w(P+2)); break;
	case  3: memmove(&M[w(P+6)], &M[w(P+4)], w(P+2)); r = 0; break;
	case  4: memset(&M[w(P+6)], w(P+4), w(P+2)); r = 0; break;
	case  5: r = memscan(w(P+6), w(P+4), w(P+2)); break;
	case  6: r = getarg(w(P+6), (char *) &M[w(P+4)], w(P+2)); break;
	case  7: r = creat((char *) &M[w(P+2)], 0644); break;
	case  8: r = t3xopen((char *) &M[w(P+4)], w(P+2)); break;
	case  9: r = close(w(P+2)); break;
	case 10: r = read(w(P+6), &M[w(P+4)], w(P+2)); break;
	case 11: r = write(w(P+6), &M[w(P+4)], w(P+2)); break;
	case 12: r = t3xseek(w(P+6), w(P+4), w(P+2)); break;
	case 13: r = rename((char *) &M[w(P+4)], (char *)&M[w(P+2)]); break;
	case 14: r = remove((char *) &M[w(P+2)]); break;
	case 15: r = t3xtrunc(w(P+2)); break;
	case 16: r = t3xbreak(w(P+2)); break;
	default: fail("bad library call"); r = 0; break;
	}
	return r & 0xffff;
}

void run(void) {
	cell	t;

	P = MEMSIZE;
	F = MEMSIZE;
	for (I = 0;; I++) {
#ifdef DEBUG
	printf("F=%04x P=%04x I=%04x %2x a=%04x A=%04x S0=%04x\n",
		F, P, I, M[I], a(), A, w(P));
#endif
	if (P < Red) fail("stack overflow");
	switch (M[I]) {
	case 0x01: push(A); break;
	case 0x02: A = 0; break;
	case 0x03: P += 2; break;
	case 0x04:
	case 0x05: A = a(); I += 2; break;
	case 0x06: A = F+a() & 0xffff; I += 2; break;
	case 0x07: A = w(a()); I += 2; break;
	case 0x08: A = w(F+a() & 0xffff); I += 2; break;
	case 0x09: Sw(a(), A); I += 2; break;
	case 0x0a: Sw(F+a() & 0xffff, A); I += 2; break;
	case 0x0b: Sw(pop(), A); break;
	case 0x0c: M[pop()] = (byte) A; break;
	case 0x0d: t = a(); Sw(t, w(t)+1 & 0xffff); I += 2; break;
	case 0x0e: t = a(); Sw(F+t & 0xffff, w(F+t & 0xffff)+1 & 0xffff);
		   I += 2; break;
	case 0x0f: A += S(a()); A &= 0xffff; I += 2; break;
	case 0x10:
	case 0x11: P += S(a()); P &= 0xffff; I += 2; break;
	case 0x12: push(P); break;
	case 0x13: Sw(a(), P); I += 2; break;
	case 0x14: A = (A<<1) + pop() & 0xffff; break;
	case 0x15: A = w(A); break;
	case 0x16: A = A + pop() & 0xffff; break;
	case 0x17: A = M[A]; break;
	case 0x18: push(I+2); I = a()-1; break;
	case 0x19: push(I); I = A-1; break;
	case 0x47:
	case 0x1a: I = a()-1; break;
	case 0x1b: I += M[I+1]+1; break;
	case 0x1c: if (0 == A) I = a()-1; else I += 2; break;
	case 0x1d: if (0 != A) I = a()-1; else I += 2; break;
	case 0x1e: if (S(pop()) >= S(A)) I = a()-1; else I += 2; break;
	case 0x1f: if (S(pop()) <= S(A)) I = a()-1; else I += 2; break;
	case 0x20: push(F); F = P; break;
	case 0x21: F = pop(); break;
	case 0x22: I = pop(); break;
	case 0x23: exit(a()); break;
	case 0x24: A = -A & 0xffff; break;
	case 0x25: A = ~A & 0xffff; break;
	case 0x26: A = 0==A? 0xffff: 0; break;
	case 0x27: A = pop() + A & 0xffff; break;
	case 0x28: A = pop() - A & 0xffff; break;
	case 0x29: A = pop() * A & 0xffff; break;
	case 0x2a: A = S(pop()) / S(A) & 0xffff; break;
	case 0x2b: A = pop() % A & 0xffff; break;
	case 0x2c: A = pop() & A; break;
	case 0x2d: A = pop() | A; break;
	case 0x2e: A = pop() ^ A; break;
	case 0x2f: A = (pop() << A) & 0xffff; break;
	case 0x30: A = (unsigned) pop() >> A; break;
	case 0x31: A = pop() == A? 0xffff: 0; break;
	case 0x32: A = pop() != A? 0xffff: 0; break;
	case 0x33: A = S(pop()) < S(A)? 0xffff: 0; break;
	case 0x34: A = S(pop()) > S(A)? 0xffff: 0; break;
	case 0x35: A = S(pop()) <= S(A)? 0xffff: 0; break;
	case 0x36: A = S(pop()) >= S(A)? 0xffff: 0; break;
	case 0x37: A = pop() * A & 0xffff; break;
	case 0x38: A = (unsigned) pop() / (unsigned) A & 0xffff; break;
	case 0x39: A = (unsigned) pop() < (unsigned) A? 0xffff: 0; break;
	case 0x3a: A = (unsigned) pop() > (unsigned) A? 0xffff: 0; break;
	case 0x3b: A = (unsigned) pop() <= (unsigned) A? 0xffff: 0; break;
	case 0x3c: A = (unsigned) pop() >= (unsigned) A? 0xffff: 0; break;
	case 0x80: A = libcall(M[I+1]); I = pop(); break;
	default:   fail("invalid opcode"); break;
	}}
}

int main(int argc, char **argv) {
	if (sizeof(int) != sizeof(int*))
		fail("must run in 32-bit mode (try compiling with -m32)");
	if (argc < 2) fail("usage: tcvm program [args]");
	Args = argv;
	Narg = argc;
	load(argv[1]);
	run();
	return EXIT_SUCCESS;
}


targets/ux386lib.c
/*
 * T3X/0 core module for generic Unix on the 386
 * Nils M Holm, 2022
 * Public Domain / 0BSD License
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>

#define BPW	4
#define cell	int

int	Argc;
char	**Argv;

cell t3x_bpw(void) { return BPW; }

char *t3x_newline(char *s) {
	s[0] = '\n';
	s[1] = 0;
	return s;
}

cell t3x_memcomp(cell n, void *d, void *s) {
	return memcmp(s, d, n);
}

cell t3x_memcopy(cell n, void *d, void *s) {
	memmove(s, d, n);
	return 0;
}

cell t3x_memfill(cell n, cell c, void *d) {
	memset(d, c, n);
	return 0;
}

cell t3x_memscan(cell n, cell c, char *s) {
	cell	i;

	for (i=0; i<n; i++)
		if (s[i] == c) return i;
	return -1;
}

cell t3x_getarg(cell k, char *s, cell n) {
	int	j, m;

	k--;
	if (n >= Argc) return -1;
	j = strlen(Argv[n]);
	m = j>=k? k: j;
	memcpy(s, Argv[n], m);
	s[m] = 0;
	return m;
}

cell t3x_create(char *s) {
	return creat(s, 0644);
}

cell t3x_open(cell mode, char *s) {
	int	r;

	if (1 == mode)
		r = creat(s, 0644);
	else if (3 == mode)
		r = open(s, O_WRONLY);
	else
		r = open(s, mode);
	if (3 == mode) lseek(r, 0L, SEEK_END);
	return r;
}

cell t3x_close(cell fd) {
	return close(fd);
}

cell t3x_read(cell k, char *b, cell fd) {
	return read(fd, b, k);
}

cell t3x_write(cell k, char *b, cell fd) {
	return write(fd, b, k);
}

cell t3x_seek(cell how, unsigned cell where, cell fd) {
	long	w;
	int	h;

	switch (how) {
		case 0:	w = where; h = SEEK_SET; break;
		case 1: w = where; h = SEEK_CUR; break;
		case 2: w = where; w = -w; h = SEEK_END; break;
		case 3: w = where; w = -w; h = SEEK_CUR; break;
		default: return 0xffff;
	}
	return lseek(fd, w, h) < 0? -1: 0;
}

cell t3x_rename(char *new, char *old) {
	return rename(old, new);
}

cell t3x_remove(char *s) {
	return remove(s);
}

cell t3x_trunc(cell fd) {
	return ftruncate(fd, lseek(fd, 0, SEEK_CUR));
}

cell	*Sem;

void handle_break(int dummy) {
	*Sem = 1;
	signal(SIGINT, handle_break);
}

cell t3x_break(cell *sem) {
	if (0 == sem) {
		signal(SIGINT, SIG_DFL);
	}
	else if ((cell *) 1 == sem) {
		/* ignore */
	}
	else {
		Sem = sem;
		*Sem = 0;
		signal(SIGINT, handle_break);
	}
	return 0;
}

void start(void);

int main(int argc, char **argv) {
	Argc = argc;
	Argv = argv;
	start();
}


targets/cpmdefs.asm
; Z80 Program Library
; Nils M Holm, 2019
; In the public domain / 0BSD license

; CP/NC system definitions, mostly identical to CP/M 2.2

WARMBOOT	equ	0	; Jump here to reboot
BDOS		equ	5	; BDOS entry point

TOPRAM		equ	BDOS+2	; First page past TPA

FCB1		equ	5Ch	; File Control Blocks
FCB2		equ	6Ch

; FCB Fields

FCB_DRIVE	equ	 0	; Drive ID, 0=default, 1=A:, ...
FCB_NAME	equ	 1	; File name
FCB_TYPE	equ	 9	; File type
FCB_EXTENT	equ	12	; Extent number
FCB_USER	equ	13	; User number
;		equ	14	; --reserved--
FCB_RECORDS	equ	15	; Number of records in this extent
FCB_BLOCKS	equ	16	; Allocation blocks used
FCB_SEQREC	equ	32	; Sequential record
FCB_RRLOW	equ	33	; Random record
FCB_RRHIGH	equ	34	; Random record, high byte
FCB_RROVFL	equ	35	; Random record, overflow

COMTAIL		equ	80h	; Command Tail

DMABUF		equ	80h	; Disk I/O (DMA) Buffer

IOBYTE		equ	0003h	; IOBYTE location

IOCONMASK	equ	 03h	; IOBYTE, Console Mask
IOAUXMASK	equ	00Ch	; IOBYTE, Auxiliary Device Mask
IOLSTMASK	equ	0C0h	; IOBYTE, List Device Mask

IOCONTTY	equ	  0	; CON: -> TTY:
IOCONCRT	equ	  1	; CON: -> CRT:
IOCONLPT	equ	  2	; CON: -> LPT: (copy)

IOAUXTTY	equ	  0	; AUX: -> TTY:
IOAUXCRT	equ	  4	; AUX: -> CRT:
IOAUXLPT	equ	  8	; AUX: -> LPT:

IOLSTTTY	equ	  0	; LST: -> TTY:
IOLSTCRT	equ	 40h	; LST: -> CRT:
IOLSTLPT	equ	 80h	; LST: -> LPT:

BDSYSRESET	equ	0	; System Reset (Warm Boot)
BDCONIN		equ	1	; Console Input, A = byte from CON:
BDCONOUT	equ	2	; Console Output, E = byte to write to CON:
BDAUXIN		equ	3	; Auxiliary Input, A = byte from AUX:
BDAUXOUT	equ	4	; Auxiliary Output, E = byte to write to AUX:
BDLSTOUT	equ	5	; List device Output, E = byte to write to LST:
BDDCONIO	equ	6	; Direct Console I/O
				; E=FF: A=byte from CON: or 0 if no input
				; E=/=FF: write to CON:
BDAUXINST	equ	7	; Get AUX: Input Status, A=FF: ready, else A=0
BDAUXOUTST	equ	8	; Get AUX: Output Status, A=FF: ready, else A=0
BDPRINTS	equ	9	; Print Console String
				; DE = '$'-terminated string
BDREADCSTR	equ	10	; Read Console String
				; DE = buffer address, first byte = maximum
				; bytes to read, second byte: bytes actually
				; read upon return
BDCONST		equ	11	; Console Status, A=0, if no incoming data
				;                 A=FF, if input ready
BDGETVER	equ	12	; Get Version, HL = version
				; H=1: MP/M, H=0: CP/M, L=0: CP/M 1.x,
				; else L=XXXXYYYY: CP/M XXXX.YYYY
BDDSKRESET	equ	13	; Disk System Reset
BDSELDSK	equ	14	; Select Disk, E=0: A:, etc
BDOPEN		equ	15	; Open File, DE=addr of FCB, A=FF: failure
BDCLOSE		equ	16	; Close File, DE=addr of FCB, A=FF: failure
BDSEARCH	equ	17	; Search First Match, DE=FCB, A=FF: failure
BDSRCHNEXT	equ	18	; Search Next Match, A=FF: failure
BDERASE		equ	19	; Erase File, DE=FCB, A=FF: failure
BDREADSEQ	equ	20	; Read Sequential, DE=FCB, A=/=0: failure
BDWRITESEQ	equ	21	; Write Sequential, DE=FCB, A=/=0: failure
BDCREATE	equ	22	; Create File, DE=FCB, A=FF: failure
BDRENAME	equ	23	; Rename File, DE=FCB, A=FF: failure
BDGETLOGVEC	equ	24	; Get Login vector (act. disks), HL = disk map
BDGETCURDSK	equ	25	; Get Current Disk, A = disk, 0=A:, etc
BDSETDMA	equ	26	; Set Disk I/O (DMA) Address, DE = addr
BDGETALVEC	equ	27	; Get Allocation Vector, HL = addr
BDSETDSKRO	equ	28	; Set Disk Read-Only, none
BDGETRODSK	equ	29	; Get Read-Only Disks, HL = disk map
BDSETFAT	equ	30	; Set File Attributes, DE=FCB, A=0: failure
BDGETDPB	equ	31	; Get Disk Parameter Block, HL = addr
BDGETSETUSR	equ	32	; Get/Set User, E=FF: get user,
				; E=/=FF: set user
BDREADRAN	equ	33	; Read Random, DE=FCB, A=0: success
BDWRITERAN	equ	34	; Write Random, DE=FCB
BDGETFSIZ	equ	35	; Get File Size, DE=FCB,
				; return in random record FCB fields
BDSETRANREC	equ	36	; Set Random Record, DE=FCB, see 35
BDRESETDSKS	equ	37	; Reset Disks, DE = disk map
BDWRITERANZ	equ	40	; Write Random with Zero-Fill, DE=FCB, see 35

BDDSKFREE	equ	46	; Get Disk Free Space, result in DMA 0-2

BDABSREAD	equ	192	; Read Absolute Record, DE = record
BDABSWRITE	equ	193	; Write Absolute Record, DE = record
BDSETDRIVE	equ	194	; Select Drive, E = drive (0 = A:, etc)
